### ISSUE TITLE
Object.constructor should be ObjectConstructor instead of Function

### ISSUE THREAD

**Issue by cervengoc** [Microsoft/TypeScript#13866]
**TypeScript Version:**  2.1.1

**Code**

```ts
type PlainObject = {
    constructor: ObjectConstructor;
    [name: string]: any
}

let o: PlainObject = { key: "value" };
```

**Expected behavior:**

It should compile as an object literal should satisfy the `PlainObject` type.

**Actual behavior:**

It gives a compile error which at the end boils down to `Object.constructor` being a `Function` and not `ObjectConstructor`.

```
Type '{ key: string; }' is not assignable to type 'PlainObject'.
  Types of property 'constructor' are incompatible.
    Type 'Function' is not assignable to type 'ObjectConstructor'.
      Property 'getPrototypeOf' is missing in type 'Function'.
```

Please see a related StackOverflow thread here: http://stackoverflow.com/a/42028363/1323504

**Comment by nitzantomer** [comment]
Also, in the [comment above](https://github.com/Microsoft/TypeScript/blob/master/lib/lib.d.ts#L99) the `Object.constructor` property it clearly states that:

> The initial value of Object.prototype.constructor is the standard built-in Object constructor

So shouldn't it be:
```
interface Object {
    /** The initial value of Object.prototype.constructor is the standard built-in Object constructor. */
    constructor: ObjectConstructor;

    ...
}
```

**Comment by mhegazy** [comment]
All other objects in the system get the `Object` properties. `C.constructor` being `ObjectConstructotr` is not correct. you need somthign like https://github.com/Microsoft/TypeScript/issues/3841 first.

**Comment by cervengoc** [comment]
@mhegazy I see. I have two followup questions.

1. Couldn't it then be at least `Function | ObjectConstructor`? I haven't thought it over yet, just a quick idea.

2. Can you suggest any workaround for implementing the above mentioned `PlainObject` stuff in current version? A `PlainObject` would like to mean roughly this:
```
function isPlainObject(obj: any) {
  return obj && obj.constructor === Object || false;
}
```

**Comment by nitzantomer** [comment]
@mhegazy 
But as the comment in the docs implies isn't the constructor an instance of `ObjectConstructor`?

**Comment by mhegazy** [comment]
> Couldn't it then be at least Function | ObjectConstructor? I haven't thought it over yet, just a quick idea.

not sure how that helps any one. `Function | ObjectConstror` is not callable nor does it have the members of `ObjectConstructor`. so kid of the worst of both worlds..

> Can you suggest any workaround for implementing the above mentioned PlainObject stuff in current version? A PlainObject would like to mean roughly this:

I am not sure i understand the scenario you are trying to achieve well enough to make any recommendations. can you elaborate.

**Comment by mhegazy** [comment]
> But as the comment in the docs implies isn't the constructor an instance of ObjectConstructor?

Not sure why would the comment say that. it just says "initial value", i.e. it can change to something else, as it is mutable.

**Comment by cervengoc** [comment]
@mhegazy About the union type, thanks for the clarification, that's what I ment about not thinking it over, sorry. 

What I would like to have is a type, which identifies plain objects. I treat an object instance as "plain", if its `constructor` equals to `Object`, so in other words it's not an instance of a class, nor an array, etc., hence the term "plain". The term itself refers to the term which `jQuery` uses in its `$.isPlainObject` implementation. 

To summarize, an instance `obj` should be instance of `PlainObject` if and only if `obj && obj.constructor === Object` is true.

Note that `jQuery`'s `isPlainObject` method is I guess much more precise on this, but for me this would be pretty much enough.

**Comment by RyanCavanaugh** [comment]
We have the new primitive type `object`, but your definition doesn't account for `Object.create(null)`:
```
> $.isPlainObject(Object.create(null))
true
> Object.create(null).constructor
undefined
> typeof Object.create(null)
"object"
```

**Comment by cervengoc** [comment]
@RyanCavanaugh thanks for the points, yes, indeed I wasn't thinking about `Object.create(null)`. I could probably refine the definition like this.

```
obj is PlainObject <=>obj && typeof obj === "object" && (obj.constructor === null || obj.constructor === Object)
```

My bigger issue with the new `object` type is that - as far as I understood from the announcement - it's compatible with any class types as well, isn't it?

**Comment by RyanCavanaugh** [comment]
> My bigger issue with the new object type is that - as far as I understood from the announcement - it's compatible with any class types as well, isn't it?

Yes

It sounds like what you want is a way to enforce "This object was created via an object literal" ? There's not a way to do that, to my knowledge.
